### ISSUE TITLE
False positive on `for` loop

### ISSUE THREAD

**Issue by egonelbre** [mgechev/revive#810]
**Describe the bug**

`for` loops where the condition has side-effects and the block is empty can be valid.

**To Reproduce**

```
package main

import "fmt"

func main() {
	next := 0
	iter := func(v *int) bool {
		*v = next
		next++
		fmt.Println(*v)
		return next < 10
	}

	z := 0
	for iter(&z) {
	}
}
```

```
$ revive .
main.go:1:1: should have a package comment
main.go:15:15: this block is empty, you can remove it
```

Such code may be present in locations where you have an iterator interface and want to skip over rest-of the entries, but still read them.

**Expected behavior**

No warning.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - go version go1.20.1 windows/amd64

**Comment by chavacava** [comment]
Hi @egonelbre, thanks for filling the issue.
Detecting if a function call has or not side effects is not simple. IMO, the cases you point are sufficiently rare to justify the (high) complexity of detecting these false positives. I see 2 solutions to the problem: disable the linter in such (rare) constructions, or if these cases are frequent in your code base, disable de whole rule.

**Comment by egonelbre** [comment]
There's also the option of warning on loops that clearly don't have side-effect. e.g. your usual `for i := 0; i < n; i++ {}` and `for _, _ = range items {}`. That way there won't be false-positives at least.

I did a quick search for such empty-blocks: https://sourcegraph.com/search?q=context:global+lang:Go+for%5C+.*%5C+%5C%7B%5Cs*%5C%7D&patternType=regexp&sm=1&groupBy=repo. (I didn't check whether revive reports warnings for all of them.)

Even in go repo has a bunch of such loops. https://sourcegraph.com/search?q=context:global+lang:Go+for%5C+.*%5C+%5C%7B%5Cs*%5C%7D+repo:github.com/golang/go&patternType=regexp&sm=1&groupBy=repo
